libs/utils/src/JobStorage.cpp

```diff
@@ -0,0 +1,228 @@
+            return nullptr;
+        }
+
+        JobId jobId = mJobOrder.front();
+        mJobOrder.pop();
+
+        it = mJobsMap.find(jobId);
+        if (it != mJobsMap.end()) {
+            break;
+        }
+
+        // If execution reaches this line, the job must have been canceled right after being added.
+        // Therefore, we should continue the loop and attempt to retrieve the next available job.
+    }
+
+    Job job = std::move(it->second);
+    mJobsMap.erase(it);
+    return job;
+}
+
+FixedCapacityVector<JobStorage::Job> JobStorage::popBatch(int const maxJobsToPop) {
+    FixedCapacityVector<Job> jobs;
+
+    if (UTILS_UNLIKELY(maxJobsToPop == 0)) {
+        return jobs;
+    }
+
+    std::lock_guard<std::mutex> lock(mStorageMutex);
+    if (mJobOrder.empty()) {
+        return jobs;
+    }
+
+    // Calculate jobs to take. If maxJobsToPop is negative, we take all jobs.
+    size_t jobsToTake = mJobOrder.size();
+    if (0 < maxJobsToPop && maxJobsToPop < jobsToTake) {
+        jobsToTake = maxJobsToPop;
+    }
+    jobs.reserve(jobsToTake);
+
+    while (0 < jobsToTake && !mJobOrder.empty()) {
+        JobId jobId = mJobOrder.front();
+        mJobOrder.pop();
+
+        auto it = mJobsMap.find(jobId);
+        if (UTILS_UNLIKELY(it == mJobsMap.end())) {
+            // The job was probably canceled.
+            continue;
+        }
+
+        jobs.push_back(std::move(it->second));
+        --jobsToTake;
+        mJobsMap.erase(it);
+    }
+
+    return jobs;
+}
+
+JobStorage::JobId JobStorage::issueJobId() noexcept {
+    std::lock_guard<std::mutex> lock(mStorageMutex);
+    JobId const jobId = mNextJobId++;
+    // Preallocate a job, which servers two main purposes. It provides a valid jobId that can be
+    // checked for integrity when passed to the `push` method, and it enables job cancellation for
+    // tasks that are yet to be pushed.
+    mJobsMap[jobId];
+    return jobId;
+}
+
+bool JobStorage::cancel(JobId const jobId) noexcept {
+    std::lock_guard<std::mutex> lock(mStorageMutex);
+
+    auto it = mJobsMap.find(jobId);
+    if (it == mJobsMap.end()) {
+        return false; // Job not found, must have been completed or canceled.
+    }
+
+    mJobsMap.erase(it);
+
+    return true;
+}
+
+void JobStorage::stop() noexcept {
+    {
+        std::lock_guard<std::mutex> lock(mStorageMutex);
+        mIsStopping = true;
+    }
+    mStorageCondition.notify_all(); // Wake up all waiting threads
+}
+
+JobWorker::~JobWorker() = default;
+
+void JobWorker::terminate() {
+    if (mStorage) {
+        mStorage->stop();
+    }
+}
+
+AmortizationWorker::AmortizationWorker(JobStorage::Ptr storage, PassKey)
+    : JobWorker(std::move(storage)) {
+}
+
+void AmortizationWorker::process(int const jobCount) {
+    if (!mStorage || jobCount == 0) {
+        return;
+    }
+
+    if (jobCount == 1) {
+        // Handle single job without vector allocation.
+        if (auto job = mStorage->pop(false)) {
+            job();
```

Comment:
Do we need to check if `job` is non-empty? I think `utils::Invocable` asserts if it's empty.

Reply:
Got it!